kratos/processes/calculate_discontinuous_distance_to_skin_process.cpp

```diff
@@ -764,6 +810,140 @@ namespace Kratos
+            return cut_edges_vector;
+        });
+
+        auto check_edge_against_distances = [&, get_neighbour_cut_edges] (const Element::GeometryType::GeometriesArrayType& rEdgeContainer, Vector& rCuteEdgeVector) {
+            for (std::size_t i_edge = 0; i_edge < mNumEdges; ++i_edge) {
+                // Edge initially cut with distances exactly 0
+                if (rCuteEdgeVector[i_edge] >= 0) {
+                    IndexType edge_i_id = std::min(rEdgeContainer[i_edge][0].Id(), rEdgeContainer[i_edge][1].Id());
+                    IndexType edge_j_id = std::max(rEdgeContainer[i_edge][0].Id(), rEdgeContainer[i_edge][1].Id());
+
+                    // Get neighbour elements in a single vector without repetitions.
+                    std::unordered_set<Element::WeakPointer, SharedPointerHasher<Element::WeakPointer>, SharedPointerComparator<Element::WeakPointer>> all_neighbour_set;
+                    for(auto& neigh : rEdgeContainer[i_edge][0].GetValue(NEIGHBOUR_ELEMENTS).GetContainer()) {
+                        all_neighbour_set.insert(neigh);
+                    }
+
+                    for(auto& neigh : rEdgeContainer[i_edge][1].GetValue(NEIGHBOUR_ELEMENTS).GetContainer()) {
+                        all_neighbour_set.insert(neigh);
+                    }
+
+                    bool is_edge_cut = false;
+                    for (auto& g_ptr_neigh : all_neighbour_set) {
+                        if (!is_edge_cut) {
+                            auto cut_edges = get_neighbour_cut_edges.Get(g_ptr_neigh);
+                            for (auto cut_edge: cut_edges) {
+                                if (cut_edge[0]==edge_i_id && cut_edge[1]==edge_j_id) {
+                                    is_edge_cut = true;
+                                    break;
+                                }
+                            }
+                        }
+                    }
+
+                    // Correcting edge if not intersected after applying an epsilon
+                    if (!is_edge_cut) {
+                        rCuteEdgeVector[i_edge] = -1;
+                    }
+                }
+            }
+        };
+
+        for (auto &r_elem : mrVolumePart.Elements())  {
+            const auto edges_container = r_elem.GetGeometry().GenerateEdges();
+            auto &r_cut_edge_vector = r_elem.GetValue(ELEMENTAL_EDGE_DISTANCES);
+            check_edge_against_distances(edges_container, r_cut_edge_vector);
+            if (mOptions.Is(CalculateDiscontinuousDistanceToSkinProcessFlags::CALCULATE_ELEMENTAL_EDGE_DISTANCES_EXTRAPOLATED))
+            {
+                auto &r_cut_edge_extra_vector = r_elem.GetValue(ELEMENTAL_EDGE_DISTANCES_EXTRAPOLATED);
+                check_edge_against_distances(edges_container, r_cut_edge_extra_vector);
+            }
+        }
+
+        KRATOS_CATCH(" ");
+    }
+
+    template <std::size_t TDim>
+    void CalculateDiscontinuousDistanceToSkinProcess<TDim>::CalculateElementalNeighbours()
```

Comment:
As far as I see, this is only used in one place. For the sake of clarity, and considering that it is a 2-lines function, I'd directly call the neighbours in there.

Furthermore, and much more important, is that you're computing the neighbours each time the distance is computed. This is a killer if the distance is to be computed at each time step or non-linear iteration. I suggest creating a `mCalculateNeighbours` which is initialized to `false` and then set to `true` after the neigbours calculation.

Reply:
Nice idea, will do it